File: CodeCrafters.FrontEnd/codecrafters-ui/src/app/features/grants/services/grant.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from '../../../core/services/api.service';
import { Observable } from 'rxjs';

export interface GrantTypeDto {
  id: string;
  code: string;
  name: string;
  purpose: string;
  fundingMinAmount: number;
  fundingMaxAmount: number;
  projectDurationMinMonths: number;
  projectDurationMaxMonths: number;
  eligibleApplicants: string;
  geographicFocus: string;
  annualCycle: string;
  maximumAwardsPerCycle: number;
  totalProgrammeBudget: number;
}

@Injectable({ providedIn: 'root' })
export class GrantService {
  constructor(private api: ApiService) {}

  getGrants(): Observable<GrantTypeDto[]> {
    return this.api.get<GrantTypeDto[]>('/grants');
  }

  getGrantById(id: string): Observable<GrantTypeDto> {
    return this.api.get<GrantTypeDto>(`/grants/${id}`);
  }
}

File: CodeCrafters.BackEnd/CodeCrafters.Infrastructure/Grants/GrantService.cs
using CodeCrafters.Application.Grants.Dtos;
using CodeCrafters.Application.Grants.Services;
using CodeCrafters.Domain.Entities;
using CodeCrafters.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace CodeCrafters.Infrastructure.Grants;

public sealed class GrantService(AppDbContext dbContext) : IGrantService
{
    public async Task<List<GrantTypeDto>> GetAllGrantsAsync(CancellationToken cancellationToken = default)
    {
        var grants = await dbContext.Set<GrantType>()
            .OrderBy(g => g.Name)
            .ToListAsync(cancellationToken);

        return grants.Select(MapToDto).ToList();
    }

    public async Task<GrantTypeDto?> GetGrantByIdAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var grant = await dbContext.Set<GrantType>()
            .FirstOrDefaultAsync(g => g.Id == id, cancellationToken);
            
        return grant is null ? null : MapToDto(grant);
    }

    private static GrantTypeDto MapToDto(GrantType grant) => new()
    {
        Id = grant.Id,
        Code = grant.Code,
        Name = grant.Name,
        Purpose = grant.Purpose,
        FundingMinAmount = grant.FundingMinAmount,
        FundingMaxAmount = grant.FundingMaxAmount,
        ProjectDurationMinMonths = grant.ProjectDurationMinMonths,
        ProjectDurationMaxMonths = grant.ProjectDurationMaxMonths,
        EligibleApplicants = grant.EligibleApplicants,
        GeographicFocus = grant.GeographicFocus,
        AnnualCycle = grant.AnnualCycle,
        MaximumAwardsPerCycle = grant.MaximumAwardsPerCycle,
        TotalProgrammeBudget = grant.TotalProgrammeBudget
    };
}
